### ISSUE TITLE
BUG: NFS mount point temporarily unlabeled with NFSv4.2 + security_label

### ISSUE THREAD

**Issue by stephensmalley** [SELinuxProject/selinux-kernel#50]
As discussed in https://lore.kernel.org/selinux/db2a9d661ec9b53e01c029f98877a3556d8297bc.camel@btinternet.com/ and https://bugzilla.redhat.com/show_bug.cgi?id=1625955, the security label of a NFSv4.2 mount point shows up with the unlabeled context for a brief period of time after mounting, and is then subsequently refreshed to the correct label.  This can yield permission denials on unlabeled_t and can break the selinux-testsuite on NFS (with not-yet-merged patches to perform tests/filesystem,fs_filesystem tests on the filesystem in which the testsuite is located rather than on separate ext4 filesystems created for the tests).  My reply to the email thread which unfortunately didn't reach the list was:

When does NFS set the security label for the root inode / mounted directory (this would be done via nfs_setsecurity() in the labeled NFS case)?  This needs to happen before it is passed to any permission calls or exposed to userspace.

For a local filesystem using xattrs (SECURITY_FS_USE_XATTR in SELinux), the initialization of the security label of the root inode happens during security_sb_set_mnt_opts() -> selinux_set_mnt_opts() -> sb_finish_set_opts() -> inode_doinit_with_dentry(root_inode, root).  In that situation, SELinux calls __vfs_getxattr() to fetch the security.selinux attribute, map it to a SID, and set it in the incore inode security structure.  However, for labeled NFS, which uses "native labeling" aka SECURITY_FS_USE_NATIVE within SELinux aka SECURITY_LSM_NATIVE_LABELS, we need NFS to pass the label into SELinux rather than having SELinux fetch an xattr since MAC labels are directly supported in the protocol and not merely extended or named attributes. This would normally happen via nfs_setsecurity() -> security_inode_notifysecctx() -> selinux_inode_notifysecctx().  However, it isn't clear to me that this occurs for the root inode prior to any further use of it by NFS.

Opening an issue here to track upstream since there hasn't been any response or fix yet from the NFS developers, the RHEL7 bug was closed wontfix, and the RHEL8 bug is not open for viewing.

**Comment by WOnder93** [comment]
I made the [RHEL-8 bug](https://bugzilla.redhat.com/show_bug.cgi?id=1660798) accessible from the public. However, there is no extra information/activity in there...

**Comment by stephensmalley** [comment]
Fixed via 28d4d0e16f09c1865e2ec8a7d89ca7057a5cf7ae
